skein/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+    N: PartialDiv<U8>,
+    <N as PartialDiv<U8>>::Output: ArrayLength<u64>,
+    N::ArrayType: Copy,
+    <<N as PartialDiv<U8>>::Output as ArrayLength<u64>>::ArrayType: Copy,
+{
+    type Output = Block<N>;
+    fn bitxor(mut self, rhs: Block<N>) -> Self::Output {
+        // XOR is endian-agnostic
+        for (s, r) in unsafe { &mut self.words }
+            .iter_mut()
+            .zip(unsafe { &rhs.words })
+        {
+            *s ^= *r;
+        }
+        self
+    }
+}
+
+#[derive(Clone)]
+struct State<X> {
+    t: (u64, u64),
+    x: X,
+}
+
+impl<X> State<X> {
+    fn new(t1: u64, x: X) -> Self {
+        let t = (0, t1);
+        State { t, x }
+    }
+}
+
+impl<X> core::fmt::Debug for State<X> {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+        f.debug_struct("State<X>")
+            .field("t", &"(unknown)")
+            .field("x", &"(unknown)")
+            .finish()
+    }
+}
+
+const VERSION: u64 = 1;
+const ID_STRING_LE: u64 = 0x3341_4853;
+const SCHEMA_VER: u64 = (VERSION << 32) | ID_STRING_LE;
+const CFG_TREE_INFO_SEQUENTIAL: u64 = 0;
+const T1_FLAG_FIRST: u64 = 1 << 62;
+const T1_FLAG_FINAL: u64 = 1 << 63;
+const T1_BLK_TYPE_CFG: u64 = 4 << 56;
+const T1_BLK_TYPE_MSG: u64 = 48 << 56;
+const T1_BLK_TYPE_OUT: u64 = 63 << 56;
+const CFG_STR_LEN: usize = 4 * 8;
+
+macro_rules! define_hasher {
+    ($name:ident, $threefish:ident, $state_bytes:ty, $state_bits:expr) => {
+        #[derive(Clone)]
+        pub struct $name<N: Unsigned + ArrayLength<u8> + NonZero + Default> {
+            state: State<Block<$state_bytes>>,
+            buffer: BlockBuffer<$state_bytes, Lazy>,
+            _output: core::marker::PhantomData<GenericArray<u8, N>>,
+        }
+
+        impl<N> core::fmt::Debug for $name<N>
+        where
+            N: Unsigned + ArrayLength<u8> + NonZero + Default,
+        {
+            fn fmt(&self, f: &mut core::fmt::Formatter) -> Result<(), core::fmt::Error> {
+                f.debug_struct("Skein").field("state", &self.state).finish()
+            }
+        }
+
+        impl<N> $name<N>
+        where
+            N: Unsigned + ArrayLength<u8> + NonZero + Default,
+        {
+            fn process_block(
+                state: &mut State<Block<$state_bytes>>,
+                block: &GenericArray<u8, $state_bytes>,
+                byte_count_add: usize,
+            ) {
+                let block = Block::from_byte_array(block);
+                state.t.0 += byte_count_add as u64;
+                let mut tweak = [0u8; 16];
+                tweak[..8].copy_from_slice(&state.t.0.to_le_bytes());
+                tweak[8..].copy_from_slice(&state.t.1.to_le_bytes());
+                let mut key = [0u8; { $state_bits / 8 }];
+                key[..].copy_from_slice(state.x.as_byte_array());
+                let fish = $threefish::new_with_tweak(&key, &tweak);
+                let mut x = block.clone();
+                fish.encrypt_block(x.as_byte_array_mut());
```

Comment:
It may be worth to tweak the `threefish` crate to make it a better fit for Skein implementation:
- Make dependency on `cipher` optional (but enabled by default). It also may help with the `minimal-versions` CI failure.
- Expose an inherent method for block encryption which works with arrays of `u64`s, something like `fn encrypt_block_u64(&self, block: &[u64; N]) -> [u64; N] -> { .. }`.
- Add variant of `new_with_tweak` which also works with arrays of `u64`s.